projectq/ops/_qubit_operator.py

```diff
@@ -226,6 +251,141 @@ def isclose(self, other, rel_tol=1e-12, abs_tol=1e-12):
                 return False
         return True
 
+    def __or__(self, qubits):
+        """
+        Operator| overload which enables the following syntax:
+
+        .. code-block:: python
+
+            QubitOperator(...) | qureg
+            QubitOperator(...) | (qureg,)
+            QubitOperator(...) | qubit
+            QubitOperator(...) | (qubit,)
+
+        Unlike other gates, this gate is only allowed to be applied to one
+        quantum register or one qubit and only if the QubitOperator is
+        unitary, i.e., consists of one term with a coefficient whose absolute
+        values is 1.
+
+        Example:
+
+        .. code-block:: python
+
+        eng = projectq.MainEngine()
+        qureg = eng.allocate_qureg(6)
+        QubitOperator('X0 X5', 1.j) | qureg  # Applies X to qubit 0 and 5
+                                             # with an additional global phase
+                                             # of 1.j
+
+        While in the above example the QubitOperator gate is applied to 6
+        qubits, it only acts non-trivially on the two qubits qureg[0] and
+        qureg[5]. Therefore, the operator| will create a new rescaled
+        QubitOperator, i.e, it sends the equivalent of the following new gate
+        to the MainEngine:
+
+        .. code-block:: python
+
+            QubitOperator('X0 X1', 1.j) | [qureg[0], qureg[5]]
+
+        which is only a two qubit gate.
+
+        Args:
+            qubits: one Qubit object, one list of Qubit objects, one Qureg
+                    object, or a tuple of the former three cases.
+
+        Raises:
+            TypeError: If QubitOperator is not unitary or applied to more than
+                       one quantum register.
+            ValueError: If quantum register does not have enough qubits
+        """
+        # Check that input is only one qureg or one qubit
+        qubits = self.make_tuple_of_qureg(qubits)
+        if len(qubits) != 1:
+            raise TypeError("Only one qubit or qureg allowed.")
+        # Check that operator is unitary
+        if not len(self.terms) == 1:
+            raise TypeError("Only unitary QubitOperators can be applied to "
```

Comment:
Only QubitOperators consisting of a single Pauli string with a coefficient of unit length can be applied.
Otherwise, people might get confused if their operator actually is unitary but consists of multiple terms.

Reply:
1/sqrt(2) * X + 1/sqrt(2) * Z = H

Reply:
Great, thanks for the example